src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java

```diff
@@ -692,93 +690,127 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
+							}
+							
+							//those of you still paying attention will note that if handled as expressions (such as by a database), these represent the same value.
+							HttpMessage msg4 = getNewMsg();
+							setParameter(msg4, param, modifiedParamValue);
 
 							try {
-								sendAndReceive(msg4Confirm, false); //do not follow redirects
+								sendAndReceive(msg4, false); //do not follow redirects
 							} catch (SocketException ex) {
 								if (log.isDebugEnabled()) log.debug("Caught " + ex.getClass().getName() + " " + ex.getMessage() + 
-										" when accessing: " + msg4Confirm.getRequestHeader().getURI().toString());
-								continue; //Something went wrong, continue to the next item in the loop
+										" when accessing: " + msg4.getRequestHeader().getURI().toString());
+								return; //Something went wrong, no point continuing
 							}
 							countExpressionBasedRequests++;
 
-							String confirmExpressionOutputUnstripped = msg4Confirm.getResponseBody().toString();
-							String confirmExpressionOutputStripped = this.stripOff(confirmExpressionOutputUnstripped, modifiedParamValueConfirm);
-
-							//set up two little arrays to ease the work of checking the unstripped output or the stripped output
-							String confirmExpressionBodyOutput[] = {confirmExpressionOutputUnstripped, confirmExpressionOutputStripped};
-
-							if (confirmExpressionBodyOutput[booleanStrippedUnstrippedIndex].compareTo(normalBodyOutput[booleanStrippedUnstrippedIndex]) != 0) {
-								//the confirm query did not return the same results.  This means that arbitrary queries are not all producing the same page output.
-								//this means the fact we earier reproduced the original page output with a modified parameter was not a coincidence
-
-								//Likely a SQL Injection. Raise it
-								String extraInfo = null;
-								if (strippedOutput[booleanStrippedUnstrippedIndex]) {
-									extraInfo = Constant.messages.getString(MESSAGE_PREFIX + "alert.expressionbased.extrainfo", modifiedParamValue, "");
-								} else {
-									extraInfo = Constant.messages.getString(MESSAGE_PREFIX + "alert.expressionbased.extrainfo", modifiedParamValue, "NOT ");
+							String modifiedExpressionOutputUnstripped = msg4.getResponseBody().toString();
+							String modifiedExpressionOutputStripped = this.stripOff(modifiedExpressionOutputUnstripped, modifiedParamValue);
+
+							//set up two little arrays to ease the work of checking the unstripped output, and then the stripped output
+							String normalBodyOutput[] = {mResBodyNormalUnstripped, mResBodyNormalStripped};
+							String expressionBodyOutput[] = {modifiedExpressionOutputUnstripped, modifiedExpressionOutputStripped};
+							boolean strippedOutput[] = {false, true};
+
+							for (int booleanStrippedUnstrippedIndex = 0; booleanStrippedUnstrippedIndex < 2 && !sqlInjectionFoundForUrl && countExpressionBasedRequests < doExpressionMaxRequests; booleanStrippedUnstrippedIndex++) {
```

Comment:
@thc, if I create the method you suggested then **countExpressionBasedRequests** will need to be passed as the parameter in that method and method has to return two things **sqlInjectionForUrl(boolean)** and new **countExpressionBasedRequests**. Method returning two variables would be a mess in JAVA.
If we can remove the loop, then it would be easy. And I don't think any usage of comparing both stripped as well stripped. I think comparing stripped will do the same job.